pkg/registry/core/service/ipallocator/controller/repair.go

```diff
@@ -156,7 +156,7 @@ func (c *Repair) doRunOnce() error {
 	snapshotByFamily := make(map[v1.IPFamily]*api.RangeAllocation)
 	storedByFamily := make(map[v1.IPFamily]ipallocator.Interface)
 
-	err := wait.PollImmediate(time.Second, 10*time.Second, func() (bool, error) {
+	err := wait.PollImmediate(time.Second, 20*time.Second, func() (bool, error) {
```

Comment:
maybe we can extend more time? 
1. Considering the case that etcd comes back available in 19s, and to execute this hook could toke a little more time.
2. Other Hook's timeout is set to `30s` or more. eg: 
https://github.com/kubernetes/kubernetes/blob/00b9ead02c37921011ebe5293558cea5277cd295/pkg/registry/rbac/rest/storage_rbac.go#L166 https://github.com/kubernetes/kubernetes/blob/a59b92e8c039fb3646dec18f9e64ee2b5462db42/pkg/registry/scheduling/rest/storage_scheduling.go#L82 https://github.com/kubernetes/kubernetes/blob/24ecb20e412cd545207a0c95cd62476f53b6d047/pkg/registry/flowcontrol/rest/storage_flowcontrol.go#L148 

WDYT, @aojea ?

Reply:
Thank you for review！Since return error will stop loop, I alse change the return blow:
`return err==nil, err  ==>  return err==nil, nil`
otherwise an error, such as `etcdserver: request timed out`, will immediately crash.

Reply:
what about printing some error log when hook fails?  since silence error make people confused.

Reply:
The test of pull-kubernetes-linter-hints reject update a deprecated function wait.PollImmediate, request use wait.PollUntilContextTimeout, I refer to other places that use PollUntilContextTimeout and use lastError to print out the error log. Please review again.

Reply:
> Considering the case that etcd comes back available in 19s, and to execute this hook also need to take some time. only leave one second remains to execute this hook may not enough though.

The timeout does not interrupt the function being executed, so it doesn't matter to use 20 seconds or 30 seconds

Reply:
> what about printing some error log when hook fails? since silence error make people confused.

Where the Fatal occuer will print fatal log. The code before has not fix this, also the code where used wait.PollUntilContextTimeout have not fix this more.  Do you have a good idea to fix this? Print log every time in loop retry?

Reply:
> > Considering the case that etcd comes back available in 19s, and to execute this hook also need to take some time. only leave one second remains to execute this hook may not enough though.
> 
> The timeout does not interrupt the function being executed, so it doesn't matter to use 20 seconds or 30 seconds

I think retrying after a timeout is more elegant than restarting apiserver directly.

Reply:
> The timeout does not interrupt the function being executed, so it doesn't matter to use 20 seconds or 30 seconds.

@aojea  
after discuss offline,  I believe that the main point of @fusida is apiserver doesn't have to restarting when `allocator.Get()` timeout, and retrying several times may be a better way. 

And I want to add a few clarifications the failure scenario here:
1. [00:00:00] first of all, etcd is down.
2. apiserver is started and trying to connect to `etcd`, with `20s` timetout setting.
3. [00:00:20] `allocator.Get()` blocks because etcd is down, until 20 seconds later it returns a timetout error.
4. [00:00:40] etcd comes back online.
5. [00:00:60] `allocator.Get()` have never be retried again, and apiserver die because of PostStartHook failed, even now etcd is available in this monent. 

And here is the reason to why a single `allocator.Get()` call timeout could lead `PostStartHook` fatil in 1min later.
1. `allocator.Get()` does block here, and it returns a `timeout error` after 20s because of timeout.
6. the outerside `wait.PollImmediate` will return `timeout error` immedially because conditionFunc returns error,  so `repair.doRunOnce` returns a error.
7. there is a `wait.Until` in  `repair.RunUntil` to make sure `repair.doRunOnce` will be executed in every 3min. so if any error happens in `repair.doRunOnce`, the next round would be 3min later.
8. however, the `PostStartHook` has a `1min` timeout, so the next round pervious mentioned will never have chance to execute. so apiserver would occurd a fatal after 1 min later.

here are about some possible fixs I thought, correct me if I'm wrong:
1. maybe we don't have to return a error when `allocator.Get()` timesout, just print a error log is enough? so `allocator.Get()` will retry immedially in next second.
2. maybe we can short the 3min interval before the first Repairing successfully? so the out side `Until` could ensure that there is another chance to run `repair.doRunOnce`.

Reply:
@xuzhenglun we are going to move to a new repair loop with the KEP1880 that will remove this codepath during the next releases, so we should try the easiest and simplest solution, it seems current approach solves the problem